core/kernel/tee_ta_manager.c

```diff
@@ -416,18 +459,26 @@ TEE_Result tee_ta_close_session(struct tee_ta_session *csess,
 	ctx = sess->ctx;
 	DMSG("Destroy session");
 
+	if (!ctx) {
+		/*
+		 * The context has been already destroyed,
+		 * free the session memory space.
+		 */
+		tee_ta_destroy_session(sess, open_sessions);
+		return TEE_SUCCESS;
+	} else if (ctx->panicked) {
+		/* Panicked TA instance, the context will be destroyed */
+		tee_ta_destroy_session(sess, open_sessions);
```

Comment:
Swap these 2 calls: first destroy the TA context referred by `sess` then destroy the `sess` reference. 

I even wonder if this case is possible.  If TA panicked before client calls `close_session`, I think its context has already been released when it panicked. Hence here i think you could `assert(!ctx->panicked);`.

Reply:
I was wonder about this scenario too and just wanted your opinion that could be obtained during review code phase. It was also the case of `tee_ta_open_session()`, but there I recoded the panic case. My thinking behind this implementation could be resumed to the cases when the TA instance get panicked from whatever reason other than through invoke command interface. Then any calls to open/close new sessions should handled this special case by not only not allowing opening a new session, but also destroying the context. The current implementation will call in `tee_ta_open_session()` in case of a panic `tee_ta_close_session()`, but there the context will not be destroyed.